Add ClusterClaim deletion validator

luolanzone opened the pull request:
1. Refine ClusterClaim webhook and add deletion validator to deny deletion
   if there is any ClusterSet referring to the ClusterClaim.
2. Move status data initialization from leader ClusterSet controller to
MemberClusterAnnounce controller and simplify the status:
* When the MemberClusterAnnounce is created, the `Ready` status will be
  `True` and `Reason` will be `Connected`.
* When the MemberClusterAnnounce is not updated in the agreed period,
  the `Ready` status will be `false` and `Reason` will be `Disconnected`.
* When the MemberClusterAnnounce is updated after Disconnected, the `Ready`
  status will be reset to `true` and `Reason` will be `Connected`, and
  `lastUpdateTime` will also be reset.
* When the MemberCLusterAnnounce is deleted, the corresponding status
  will be removed.

Signed-off-by: Lan Luo <luola@vmware.com>

luolanzone commented:
/test-multicluster-e2e

luolanzone commented on multicluster/controllers/multicluster/commonarea/remote_common_area.go:
```diff
@@ -279,27 +277,15 @@ func (r *remoteCommonArea) updateRemoteCommonAreaStatus(connected bool, err erro
 	defer r.mutex.Unlock()
 	r.mutex.Lock()
 
-	if leaderID == common.InvalidClusterID {
```
The `InvalidClusterID` will be set only when the ClusterSet is not created and common area is nil or the ClusterSet is deleted and common area is reset to nil. So there is no case to call `updateLeaderStatus` when the leader ID is `InvalidClusterID`.

luolanzone commented:
/test-multicluster-e2e

jianjuns commented on multicluster/cmd/multicluster-controller/clusterclaim_webhook.go:
```diff
@@ -0,0 +1,106 @@
+		if len(clusterSetList.Items) > 0 {
+			// ClusterSet webhook will guarantee that there is only one ClusterSet at most in a given Namespace.
+			existingClusterSet = clusterSetList.Items[0]
+			klog.InfoS("value", "value", clusterClaim.Value, "clsutersetname", existingClusterSet.Name, "req.name", req.Name)
```
Debug log?

luolanzone replied:
Removed

jianjuns commented on multicluster/cmd/multicluster-controller/clusterclaim_webhook.go:
```diff
@@ -0,0 +1,106 @@
+		}
+		// Deny ClusterClaim deletion if the ClusterClaim is referred in a ClusterSet.
+		if deny {
+			klog.ErrorS(err, "Error deleting ClusterClaim, it's referred by a ClusterSet", "ClusterClaim", klog.KObj(clusterClaim), "ClusterSet", klog.KObj(&existingClusterSet))
```
"The ClusterClaim is referred by a ClusterSet. Cannot delete it"

jianjuns replied:
I feel better to log only the Claim/Set names, rather than the whole object which is quite unnecessary.

luolanzone replied:
Message updated. `klog.KObj` only returns an object contains Namespace and Name, so the purpose of `klog.KObj` is to print Name and Namespace, not the whole object.

jianjuns commented on multicluster/cmd/multicluster-controller/clusterclaim_webhook.go:
```diff
@@ -0,0 +1,106 @@
+		// Deny ClusterClaim deletion if the ClusterClaim is referred in a ClusterSet.
+		if deny {
+			klog.ErrorS(err, "Error deleting ClusterClaim, it's referred by a ClusterSet", "ClusterClaim", klog.KObj(clusterClaim), "ClusterSet", klog.KObj(&existingClusterSet))
+			return admission.Denied(fmt.Sprintf("the ClusterClaim %s is referred by a ClusterSet %s, please delete the ClusterSet first\n", klog.KObj(clusterClaim), klog.KObj(&existingClusterSet)))
```
Ditto

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -114,28 +110,26 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
-					data.leaderStatus.reason = ReasonConnectedLeader
-				}
+			if !apierrors.IsNotFound(err) {
+				return ctrl.Result{}, err
```
We need to unlock. I suggest to set err to nil if apierrors.IsNotFound(err), and break and check err and return after line 135.

luolanzone replied:
Fixed, defined a variable `err`.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -114,28 +110,26 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
-			// Nothing to handle in this case and MemberClusterAnnounce will also be deleted soon.
-			// TODO: Add ClusterClaim webhook to make sure it cannot be deleted while ClusterSet is present.
+			// If it's not found error, ClusterClaims were probably deleted during the processing of MemberClusterAnnounce.
+			// Nothing to handle in this case since it means no ClusterSet is referring to it or the ClusterSet is deleted
```
What does "no ClusterSet is referring to it" mean?

luolanzone replied:
I mixed the wrong message. It should say 'since the ClusterSet is deleted'.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -114,28 +110,26 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
-			// TODO: Add ClusterClaim webhook to make sure it cannot be deleted while ClusterSet is present.
+			// If it's not found error, ClusterClaims were probably deleted during the processing of MemberClusterAnnounce.
+			// Nothing to handle in this case since it means no ClusterSet is referring to it or the ClusterSet is deleted
+			// as well, no status need to be updated.
```
"and no need to update cluster status"

I actually suggest to remove this sentence.

luolanzone replied:
Removed the whole sentence.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -114,28 +110,26 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
+		// Check whether this local cluster is the leader for this member.
+		clusterClaim := &multiclusterv1alpha2.ClusterClaim{}
+		if err := r.Get(context.TODO(), types.NamespacedName{Name: multiclusterv1alpha2.WellKnownClusterClaimID, Namespace: req.Namespace}, clusterClaim); err == nil {
+			if clusterClaim.Value == memberAnnounce.LeaderClusterID {
```
When possible the leader  clusterID does not match? Again, I suggest to remove laderClusterID from memberAnnounce.

luolanzone replied:
It's possible that user creates a member ClusterSet with right access but incorrect leader ClusterID, if we didn't validate the leader ID and allow it to create the member cluster announce, then member cluster will show a status with wrong cluster ID but connected status.

jianjuns replied:
I actually feel leader cluster ID in ClusterSet is not very useful. It is not used anywhere, except these validations. But anyway, we can keep the current way, if you prefer.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -114,28 +110,26 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
-			// If err != nil, probably ClusterClaims were deleted during the processing of MemberClusterAnnounce.
-			// Nothing to handle in this case and MemberClusterAnnounce will also be deleted soon.
-			// TODO: Add ClusterClaim webhook to make sure it cannot be deleted while ClusterSet is present.
+			// If it's not found error, ClusterClaims were probably deleted during the processing of MemberClusterAnnounce.
```
I feel we should revisit the design of Add/RemoveMember() after ClusterSet changes. Why that indirect, but not start/stop memberAnnounce periodic check right after the memberAnnounce is created/deleted in MemberClusterAnnounceReconciler?

Pasted below my earlier comments too.
> 1. Do we really need leader cluster ID in the MemberAnnounce?
> 2. Do we still need all status conditions after we remove election?
> 3. Add/RemoveMember() need not to be called by ClusterSet changes, after we change to auto-update ClusterSet, but we can start to maintain the timerdata once MemberClusterAnnounce is created?

luolanzone replied:
Yeah, from member cluster's perspective, I feel we don't have to provide leader's ClusterID as long as the access token is correct. We probably need to refactor the whole ClusterSet part including the ClusterSet spec vs ClusterSet status for joined member.

jianjuns replied:
We can decide leader ClusterID later. But please consider 3). The current code is unnecessarily complex.

luolanzone replied:
Sure, I will double check this part.

luolanzone replied:
Hi @jianjuns  I have refined this part, the MemberClusterAnnounce is responsible to initialize the data now. Could you help to take another look? thanks.

luolanzone commented on multicluster/cmd/multicluster-controller/clusterclaim_webhook.go:
```diff
@@ -0,0 +1,106 @@
+		deny := false
+		var existingClusterSet mcv1alpha1.ClusterSet
+		if len(clusterSetList.Items) > 0 {
+			// ClusterSet webhook will guarantee that there is only one ClusterSet at most in a given Namespace.
```
Done

luolanzone commented on multicluster/cmd/multicluster-controller/clusterclaim_webhook.go:
```diff
@@ -0,0 +1,106 @@
+				return admission.Errored(http.StatusBadRequest, err)
+			}
+			if oldClusterClaim.Value != clusterClaim.Value {
+				klog.ErrorS(err, "the field 'value' is immutable", "ClusterClaim", klog.KObj(clusterClaim))
```
Fixed

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -114,38 +108,34 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
 		klog.V(2).InfoS("Reset lastUpdateTime", "cluster", memberAnnounce.ClusterID)
```
After reading the code more, I wonder should we do the following two:

1. Merge timerData and memberStatus to one map (and call it memberStatusMap)
2. Move this block to addMember() (and rename the func to sth like addUpdateMemberStatus())

What you think?

luolanzone replied:
Yes, I am thinking it should work, but If we'd like to include this PR to 1.8, I feel it's better to has a follow up PR to rethink about the whole status part, what's your suggestion?

jianjuns replied:
I am fine to do another PR, but I do feel we should clean up the code. The current implementation is unnecessarily complex, hard to follow and may include bugs due to that.

luolanzone replied:
Sure, I feel we need rethink the design including ClusterClaim/ClusterSet and status.

luolanzone commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -114,38 +108,34 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
 		klog.V(2).InfoS("Reset lastUpdateTime", "cluster", memberAnnounce.ClusterID)
```
Done, changed it to `If the member is not found in timerData, it means MemberClusterAnnounce is not created
or it was deleted.`

luolanzone commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -97,14 +95,10 @@ func NewMemberClusterAnnounceReconciler(client client.Client, scheme *runtime.Sc
-			// Cannot read the requested resource. Return error, so reconciliation will be retried.
-			return ctrl.Result{}, err
-		}
-		// If MemberClusterAnnounce is deleted, no need to process because RemoveMember must already
```
Done

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -114,38 +108,34 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
 	}
 	r.mapLock.Unlock()
 
+	if err != nil && !apierrors.IsNotFound(err) {
```
I meant if you do not break the loop for an error, "err" can be overridden to "nil" later.

luolanzone replied:
Sorry, I didn't get the point here, which loop do you refer to?

jianjuns replied:
Sorry, my mistake here. I thought this line is in a loop (probably I mistook the block with some other code as I did not see the whole block from the review page).

Then there is no issue.

luolanzone committed:
Add ClusterClaim deletion validator

1. Refine ClusterClaim webhook, add deletion validator to deny deletion
   if there is any ClusterSet referring to it.
2. Move status data initialization from leader ClusterSet controller to
MemberClusterAnnounce controller and simplify the status:
* When the MemberClusterAnnounce is created, the `Ready` status will be
  `True` and `Reason` will be `Connected`.
* When the MemberClusterAnnounce is not updated in the agreed period,
  the `Ready` status will be `false` and `Reason` will be `Disconnected`.
* When the MemberClusterAnnounce is updated after Disconnected, the `Ready`
  status will be reset to `true` and `Reason` will be `Connected`,
  lastUpdateTime will also be reset.
* When the MemberCLusterAnnounce is deleted, the corresponding status
  will be removed.

Signed-off-by: Lan Luo <luola@vmware.com>

jianjuns commented on multicluster/cmd/multicluster-controller/clusterclaim_webhook.go:
```diff
@@ -0,0 +1,105 @@
+		// Deny ClusterClaim deletion if the ClusterClaim is referred in a ClusterSet.
+		if deny {
+			klog.ErrorS(err, "The ClusterClaim is referred by a ClusterSet. Cannot delete it", "ClusterClaim", klog.KObj(clusterClaim), "ClusterSet", klog.KObj(&existingClusterSet))
+			return admission.Denied(fmt.Sprintf("the ClusterClaim %s is referred by a ClusterSet %s, please delete the ClusterSet first\n", klog.KObj(clusterClaim), klog.KObj(&existingClusterSet)))
```
, please -> Please

jianjuns commented on multicluster/cmd/multicluster-controller/clusterclaim_webhook.go:
```diff
@@ -0,0 +1,105 @@
+	switch req.Operation {
+	case admissionv1.Create:
+		if clusterClaim.Name != mcv1alpha2.WellKnownClusterClaimClusterSet && clusterClaim.Name != mcv1alpha2.WellKnownClusterClaimID {
+			return admission.Denied(fmt.Sprintf("name %s is not valid, only 'id.k8s.io' and 'clusterset.k8s.io' are valid names for ClusterClaim\n", clusterClaim.Name))
```
, only -> . Only

jianjuns commented on multicluster/controllers/multicluster/commonarea/remote_common_area.go:
```diff
@@ -239,21 +238,20 @@ func (r *remoteCommonArea) SendMemberAnnounce() error {
+		return nil
 	}
 
+	// Create happens first before the leader validation pass. When the creation is successful,
```
pass -> passes

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -69,23 +58,18 @@ type MemberClusterAnnounceReconciler struct {
 	mapLock      sync.RWMutex
-	memberStatus map[common.ClusterID]*multiclusterv1alpha1.ClusterStatus
-	timerData    map[common.ClusterID]*timerData
+	memberStatus map[common.ClusterID]*memberData
```
Suggest to rename it memberStatusMap or memberStatuses

luolanzone replied:
Done

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -97,55 +81,50 @@ func NewMemberClusterAnnounceReconciler(client client.Client, scheme *runtime.Sc
 	r.mapLock.Lock()
-	// If the member is not found in timerData, it means the member is not added to the ClusterSet or was deleted.
-	if data, ok := r.timerData[common.ClusterID(memberAnnounce.ClusterID)]; ok {
+	// If the member is not found in memberStatus, it means MemberClusterAnnounce is not created
```
Again, I think this block (from line 94 to 119) should be moved to addMemberStatus (probably rename it to addOrUpdateMemberStatus). Could you check?

luolanzone replied:
Updated.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -163,6 +142,7 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
 		klog.ErrorS(err, "Failed to add member cluster to ClusterSet", "cluster", memberAnnounce.ClusterID)
 		return ctrl.Result{}, err
 	}
+	r.addMemberStatus(memberID, v1.ConditionTrue)
```
Should we move it to the end of the func after adding the Finalizer?

luolanzone replied:
I think it's better to place it before adding Finalizer since member is actually connected and able to do export/import etc.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -163,6 +142,7 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
 		klog.ErrorS(err, "Failed to add member cluster to ClusterSet", "cluster", memberAnnounce.ClusterID)
```
We should ignore "already exists" error in addMemberToClusterSet (in my mind better to add comments on addMemberToClusterSet to explain the behavior)?

luolanzone replied:
Yes, return nil for `already exists` error.

jianjuns replied:
Oh, I did not see that. Then we should move r.addOrUpdateMemberStatus() to line 113 before "if common.StringExistsInSlice(memberAnnounce.Finalizers, finalizer)"?

I assume we can call the func only once, and remove the call at line 95.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -163,6 +142,7 @@ func (r *MemberClusterAnnounceReconciler) Reconcile(ctx context.Context, req ctr
 		return ctrl.Result{}, err
 	}
+	r.addMemberStatus(memberID, v1.ConditionTrue)
 	klog.InfoS("Adding finalizer to MemberClusterAnnounce", "MemberClusterAnnounce", klog.KObj(memberAnnounce))
```
We will repeatedly log this message for every memberAnnounce update. Should should check if Finalizer is already added?

luolanzone replied:
When the finalizer exists, it will return, code won't reach here.
```
	if common.StringExistsInSlice(memberAnnounce.Finalizers, finalizer) {
		return ctrl.Result{}, nil
	}
```

luolanzone commented:
/test-multicluster-e2e

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -97,55 +80,27 @@ func NewMemberClusterAnnounceReconciler(client client.Client, scheme *runtime.Sc
+	// Initialize the member status with unknown status when controller restarts.
+	// Let's background process to check and update the connection status.
+	// v1.ConditionUnknown will be ignored if status already exists in the map.
+	r.addOrUpdateMemberStatus(memberID, v1.ConditionUnknown)
```
Can we remove this call, and keep only a single call of addOrUpdateMemberStatus(), e.g. at line 112?

luolanzone replied:
Fixed.

jianjuns commented on multicluster/controllers/multicluster/commonarea/remote_common_area.go:
```diff
@@ -206,29 +207,27 @@ func getSecretCACrtAndToken(secretObj *v1.Secret) ([]byte, []byte, error) {
-			}
-		}
+
+	if apierrors.IsNotFound(err) {
```
We can just do:

localClusterMemberAnnounceExists := err == nil

luolanzone replied:
Done

jianjuns commented on multicluster/controllers/multicluster/commonarea/remote_common_area.go:
```diff
@@ -206,29 +207,27 @@ func getSecretCACrtAndToken(secretObj *v1.Secret) ([]byte, []byte, error) {
+	}, existingMemberAnnounce); err != nil && !apierrors.IsNotFound(err) {
 		return err
 	}
+	var localClusterMemberAnnounceExist bool
```
localClusterMemberAnnounceExist -> localClusterMemberAnnounceExists

luolanzone replied:
Done

jianjuns commented on multicluster/controllers/multicluster/commonarea/remote_common_area.go:
```diff
@@ -364,7 +350,7 @@ func (r *remoteCommonArea) Start() context.CancelFunc {
 func (r *remoteCommonArea) doMemberAnnounce() {
 	if err := r.SendMemberAnnounce(); err != nil {
-		klog.ErrorS(err, "Error writing member announce", "cluster", r.GetClusterID())
+		klog.ErrorS(err, "Error writing MemberClusterAnnounce", "cluster", r.GetClusterID())
```
writing -> updating

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -97,55 +80,22 @@ func NewMemberClusterAnnounceReconciler(client client.Client, scheme *runtime.Sc
-	err := r.Get(ctx, req.NamespacedName, memberAnnounce)
+	var err error
+	memberID := getIDFromName(req.Name)
+	err = r.Get(ctx, req.NamespacedName, memberAnnounce)
```
Nit: you can keep the previous version - err := r.Get(ctx, req.NamespacedName, memberAnnounce)

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -97,55 +80,22 @@ func NewMemberClusterAnnounceReconciler(client client.Client, scheme *runtime.Sc
 			klog.ErrorS(err, "Failed to remove member cluster from the ClusterSet", "cluster", memberAnnounce.ClusterID)
 			return ctrl.Result{}, err
 		}
+		r.removeMemberStatus(memberID)
```
Question - should we call removeMemberStatus() after removing Finalizer? E.g. is it possible that a member cluster can clean the DeletionTimestamp and continue to update/reuse the Announce?

luolanzone replied:
From member's perspective, the MemberClusterAnnounce will be deleted when ClusterSet is deleted, so it shouldn't be able to talk to leader any more.

jianjuns replied:
>  E.g. is it possible that a member cluster can clean the DeletionTimestamp and continue to update/reuse the Announce?

Even a member cluster should never do that, could a user manually do that?

luolanzone replied:
I don't think DeletionTimestamp is mutable, here is the comment of `DeletionTimestamp`.
```
// Once the deletionTimestamp is set, this value may not be unset or be set further into the
// future, although it may be shortened or the resource may be deleted prior to this time.
```

jianjuns replied:
Ok.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -256,61 +185,45 @@ func (r *MemberClusterAnnounceReconciler) processMCSStatus() {
 	}
-
-	conditions := make([]multiclusterv1alpha1.ClusterCondition, 0, 2)
+	conditions := make([]multiclusterv1alpha1.ClusterCondition, 0, 1)
```
Nit: maybe better to add a new line.

jianjuns commented on multicluster/controllers/multicluster/memberclusterannounce_controller.go:
```diff
@@ -256,61 +185,45 @@ func (r *MemberClusterAnnounceReconciler) processMCSStatus() {
-		LastTransitionTime: metav1.Now(),
-		Message:            "Member created",
-		Reason:             ReasonNeverConnected,
+		Message:            "Member Created",
```
Question - what is difference between  "Member Created" and "Member Connected"? Should we just set the condition to "ConditionTrue + Member Connected"? Or there is a protocol that member should notify leader after it is ready by updating the Announce?

luolanzone replied:
`"Member Created"` will be used only when controller starts.  It will become `Member Connected"` when member cluster send a MemberClusterAnnounce update. The purpose is handle the case `MemberClusterAnnounce` aleady exists on leader cluster, but it might be stale and will be cleaned up soon.

jianjuns replied:
But I meant what is the problem we set the condition to "ConditionTrue + Member Connected" directly here? If the member cluster does not update the Announce, it will be changed to "Disconnected" anyway?

luolanzone replied:
Yes, it will be changed to "Disconnected", I feel both way works. Do you recommend to use `Member Connected` only?

luolanzone replied:
It will be updated every 5 seconds if member cluster works well. But you are right, I feel stale MemberClusterAnnounce case will be rare, let me change it to "ConditionTrue + Member Connected".

jianjuns replied:
BTW, do you think 5 seconds may be too frequent?

luolanzone replied:
Yeah, I feel  it can be longer, how about 15s?

jianjuns replied:
Was thinking about 10 seconds. 15 seconds sound good to me too.

luolanzone replied:
OK, I updated it to 10s in this PR, it can be quicker to detect failure.

jianjuns reviewed (commented):
Questions and a few nits.

luolanzone commented:
/test-multicluster-e2e

jianjuns commented on multicluster/controllers/multicluster/commonarea/remote_common_area.go:
```diff
@@ -206,29 +207,21 @@ func getSecretCACrtAndToken(secretObj *v1.Secret) ([]byte, []byte, error) {
-			localClusterMemberAnnounce.LeaderClusterID = string(leaderID)
-		}
 
+	localClusterMemberAnnounceExists := !apierrors.IsNotFound(err)
```
Is it safe to pass nil to apierrors.IsNotFound()? I would suggest:

localClusterMemberAnnounceExists := err == nil

luolanzone replied:
Done

jianjuns commented:
/test-multicluster-e2e
/test-integration

jianjuns commented:
/skip-all
